### PR TITLE
prometheus and grafana dashboards implemented

### PR DESCRIPTION
**Pull request by pavo-tusker**
Implemented Basic Prometheus Metrics and Grafana Dashboard using this [Prometheus Crate](https://crates.io/crates/prometheus) [#496](https://github.com/meilisearch/product/issues/496)
Screenshot from 2022-08-04 19-59-06
Screenshot from 2022-08-04 19-58-56
Screenshot from 2022-08-04 19-58-40

### CODE REVIEW

**Comment by pavo-tusker** [comment]
@irevoire Fixed the issues in the review.

**Comment by irevoire** [comment]
Hey, I'm off today I'll look into it tomorrow!
But thanks a lot for your contribution it looks sick 🤘

File: meilisearch-http/src/routes/mod.rs
Comment (irevoire) [inline meilisearch-http/src/routes/mod.rs:283]:
```suggestion
```

File: meilisearch-http/src/lib.rs
Comment (irevoire) [inline meilisearch-http/src/lib.rs:183]: From what I see/understand, Prometheus is using [`Atomic`](https://docs.rs/prometheus/latest/src/prometheus/atomic64.rs.html) to update its state.
Most access are relaxed, but the `inc` is an `Acquire` for example (which can be slower than a mutex iirc).

And this code is executed for EVERY call to ANY route.

Maybe it could slow down mielisearch when it’s under high pressure?
What do you think @meilisearch/core-team?

Personally, I still think we should accept this PR, my question is, should it work like that, or should we put the feature behind a cli args so it only slows down those who need prometheus?

File: meilisearch-http/src/routes/mod.rs
Comment (irevoire) [inline meilisearch-http/src/routes/mod.rs:299]:
```suggestion
```

We’re not using this buffer later, we can let it be dropped later

File: meilisearch-http/src/routes/mod.rs
Comment (irevoire) [inline meilisearch-http/src/routes/mod.rs:298]:
```suggestion
    let response = String::from_utf8(buffer).expect("Failed to convert bytes to string");
```
Not sure but since we don’t need the buffer later we can simply give it to the string without cloning it maybe?

**Review by irevoire — CHANGES_REQUESTED** [review]
Hey, @pavo-tusker thank you for your contribution it’s awesome, I want to see what the rest of the @meilisearch/core-team thinks of it before merging it.

But overall great contribution, thanks!

File: meilisearch-http/src/lib.rs
Comment (ManyTheFish) [reply meilisearch-http/src/lib.rs:183]: Good question @irevoire, I agree with you that this PR is a great improvement.

I like the idea of setting an opt-in flag for this feature as we made for the auto-batching beta feature.
Moreover, this feature didn't pass the usual feature process and didn't have any approval from the product team.

So, I'd find it safer to deactivate by default this feature. 

But again, it's a great feature that could be useful for some of our users and for us. 😊

Comment (davelarkan) [reply meilisearch-http/src/lib.rs:183]: This PR looks great and would be incredibly helpful for us on the cloud side.

Having this be an opt-in feature would be fine (we would likely enable it by default for all projects)

Comment (gmourier) [reply meilisearch-http/src/lib.rs:183]: @ManyTheFish It seems it could be a strong basis for a more [detailed exporter](https://github.com/meilisearch/product/issues/496) that is considered for the future.

**Comment by irevoire** [comment]
Ok @pavo-tusker, so after chatting with the team we would really like to integrate your change but with the option to disable it.
Could you create an `--enable-metrics-route` cli option as @ManyTheFish said that disables both the `metrics` route and the hook in actix (=the `wrap_fn`).

**Comment by irevoire** [comment]
Hey, just so you know, [clippy was updated](https://github.com/meilisearch/meilisearch/pull/2665) last weekend thus you'll need to rebase on main to make the CI pass again.

**Comment by pavo-tusker** [comment]
> Hey, just so you know, [clippy was updated](https://github.com/meilisearch/meilisearch/pull/2665) last weekend thus you'll need to rebase on main to make the CI pass again.

Hey, @irevoire  Just added the " --enable-metrics-route " option as per the above discussions. By default, metrics will be disabled. passing "--enable-metrics-route" arg will enable /metrics route and Req/Res metrics. Removed wrap_fn and changed it as a conditional wrap middleware. 

and what do I need to do on Clippy? can you give some more details about this?

File: meilisearch-http/src/option.rs
Comment (irevoire) [inline meilisearch-http/src/option.rs:150]:
```suggestion
    /// Enables Prometheus metrics and /metrics route.
    #[clap(long)]
```

**Review by irevoire — CHANGES_REQUESTED** [review]
Hey @pavo-tusker we're almost there!

After accepting my last change you will need to rebase on main (there should be no conflict from what I see) and I’ll be able to merge your change!
It should be something like that depending on your workflow.
```
git checkout main
git pull upstream/main
git checkout -
git rebase main
git push -f
```

**Comment by irevoire** [comment]
bors merge

File: meilisearch-http/src/option.rs
Comment (pavo-tusker) [reply meilisearch-http/src/option.rs]: Hey, I did rebase the main to the feature branch. Still,
Screenshot from 2022-08-17 21-09-57
 the Clippy is failing.

Comment (irevoire) [reply meilisearch-http/src/option.rs]: It's super strange; I can't reproduce locally.
But if you can go on, just fix it here and make it stop.

If you're not familiar with Clippy;
You can run `cargo clippy --fix` to automatically fix most of the warnings.

Then don't forget to run `cargo fmt` and you should be good!

**Review by irevoire — DISMISSED** [review]
Thanks for your contribution!
bors merge

**Comment by pavo-tusker** [comment]
> Thanks for your contribution! bors merge

Again there is an issue with clippy :( on new bors merged code.
Just fixed that too.

**Review by irevoire — APPROVED** [review]
yeah I don't know what's happening, I'll try again and if it doesn't work we're probably going to bypass bors or remove clappy from the CI 🤔 
bors merge

**Comment by irevoire** [comment]
Hey @pavo-tusker, finally, we decided we wanted to take more time to refine the feature, thus for the next release (happening today), we're going to move the metrics behind a feature flag (which requires you to build your own meilisearch binary in order to use it).

For the next weeks, we’re going to heavily use it and see what other information we can provide, and if, in the end, we are satisfied with the outcome, we’ll probably make it official for the v1.0 (in 4 months) :tada: 
We’ll see how it goes, but I can't promise anything.

I would love to hear your use case, though; why did you feel the need to implement this route, and is it ok for you to recompile meilisearch from scratch in the meantime?

**Comment by pavo-tusker** [comment]
> Ease observability meilisearch/product#496

Sure, 🫡. I was exploring meilisearch for our needs, and observability is something I felt helped understand the queries we did, so I added the same and thought I would make a  pull request to contribute. Using as a binary now.
Would be great to see the merge when it happens, let me know if anything is needed from me.

**Pull request merged**
